pkg/scheduler/framework/cycle_state_test.go

```diff
@@ -31,44 +34,92 @@ func (f *fakeData) Clone() StateData {
+// createCycleStateWithFakeData creates *CycleState with fakeData.
+// The given data is used in stored fakeData.
+func createCycleStateWithFakeData(data string, recordPluginMetrics bool) *CycleState {
+	c := NewCycleState()
+	c.Write(key, &fakeData{
+		data: data,
+	})
+	c.SetRecordPluginMetrics(recordPluginMetrics)
+	return c
+}
 
-	valueCopy, err := stateCopy.Read(key)
-	if err != nil {
-		t.Errorf("failed to read copied value: %v", err)
-	}
-	if v, ok := valueCopy.(*fakeData); ok && v.data != data1 {
-		t.Errorf("clone failed, got %q, expected %q", v.data, data1)
+func TestCycleStateClone(t *testing.T) {
+	tests := []struct {
+		name            string
+		state           *CycleState
+		wantClonedState *CycleState // clonedState is CycleState cloned by Clone().
+	}{
+		{
+			name:            "clone with recordPluginMetrics true",
+			state:           createCycleStateWithFakeData("data", true),
+			wantClonedState: createCycleStateWithFakeData("data", true),
+		},
+		{
+			name:            "clone with recordPluginMetrics false",
+			state:           createCycleStateWithFakeData("data", false),
+			wantClonedState: createCycleStateWithFakeData("data", false),
+		},
+		{
+			name:            "clone with nil CycleState",
+			state:           nil,
+			wantClonedState: nil,
+		},
 	}
 
-	originalValue.data = data2
-	original, err := state.Read(key)
-	if err != nil {
-		t.Errorf("failed to read original value: %v", err)
-	}
-	if v, ok := original.(*fakeData); ok && v.data != data2 {
-		t.Errorf("original value should change, got %q, expected %q", v.data, data2)
-	}
+	cmpOpts := []cmp.Option{
+		cmp.AllowUnexported(CycleState{}),
+		cmp.Comparer(func(a, b sync.Map) bool {
+			isEqual := true
+
+			countA := 0
+			a.Range(func(k, v1 interface{}) bool {
+				countA++
+				v2, ok := b.Load(k)
+				if !ok {
+					isEqual = false
+					return false
+				}
+
+				if !cmp.Equal(v1, v2, cmp.AllowUnexported(fakeData{})) {
+					isEqual = false
+					return false
+				}
+				return true
+			})
+
+			countB := 0
+			b.Range(func(k, _ interface{}) bool {
+				countB++
+				return true
+			})
 
-	if v, ok := valueCopy.(*fakeData); ok && v.data != data1 {
-		t.Errorf("cloned copy should not change, got %q, expected %q", v.data, data1)
+			if countA != countB {
+				return false
+			}
+			return isEqual
+		}),
 	}
-}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			state := tt.state
+			stateCopy := state.Clone()
+
+			if diff := cmp.Diff(state, stateCopy, cmpOpts...); diff != "" {
+				t.Errorf("Unexpected cloned state (-want, +got):\n%s", diff)
+			}
+
+			if tt.state == nil {
+				return
+			}
 
-func TestCycleStateCloneNil(t *testing.T) {
-	var state *CycleState
-	stateCopy := state.Clone()
-	if stateCopy != nil {
-		t.Errorf("clone expected to be nil")
+			stateCopy.Write(key, &fakeData{data: "modified"})
+			if diff := cmp.Diff(state, stateCopy, cmpOpts...); diff == "" {
+				t.Errorf("Unexpected cloned state (-want, +got):\n%s", diff)
```

Comment:
This message is weird. We want an error if they are equal, in which case `diff == ""`. Maybe just use `cmp.Equal` in this case?

Reply:
Fixed the error message and change to use `cmp.Equal`